packages/wonder-blocks-form/src/hooks/use-field-validation.ts

```diff
@@ -0,0 +1,146 @@
+ * - `onChangeValidation` - Validation logic for when a field changes.
+ *
+ * @example
+ *  export const MyComponent = ({
+ *     value,
+ *     disabled,
+ *     validate,
+ *     onValidate,
+ *     required,
+ *     instantValidation,
+ *  }) => {
+ *   const {errorMessage, onBlurValidation, onChangeValidation} =
+ *       useFieldValidation({
+ *           value,
+ *           disabled,
+ *           validate,
+ *           onValidate,
+ *           required,
+ *           instantValidation,
+ *       });
+ *     return (
+ *       <input
+ *           onBlur={(event) => {
+ *               onBlurValidation(event.target.value);
+ *           }}
+ *           onChange={(event) => {
+ *               onChangeValidation(event.target.value);
+ *           }}
+ *           aria-invalid={!!errorMessage}
+ *       />
+ *   );
+ * }
+ *
+ */
+export const useFieldValidation = ({
+    value,
+    disabled = false,
+    validate,
+    onValidate,
+    required = false,
+    instantValidation = true,
+}: FieldValidationProps) => {
+    const [errorMessage, setErrorMessage] = React.useState<string | null>(
+        // Ensures error is updated on unmounted server-side renders
+        // Pass in an initializer function so the validate prop is not called
+        // on every render here
+        () =>
+            (validate && value !== "" && !disabled && validate(value)) || null,
+    );
+
+    const onChangeValidation = (newValue: string) => {
+        if (instantValidation) {
+            handleValidation(newValue);
+        } else {
+            if (errorMessage) {
+                // If instantValidation is false and there is an error
+                // message, error needs to be cleared when the user updates
+                // the value
+                setErrorMessage(null);
+                if (onValidate) {
+                    onValidate(null);
+                }
+            }
+        }
+    };
+
+    const onBlurValidation = (newValue: string) => {
+        // Only handle validation on blur if instantValidation is false
+        if (!instantValidation) {
+            // Handle validation on blur if:
+            // 1. There is a value in the field so a user tabbing through
+            // fields doesn't trigger the error state when the field is left
+            // empty. Or,
+            // 2. The field is required. Tabbing through an empty field that
+            // is required will trigger the error state
+            if (newValue || required) {
+                handleValidation(newValue);
+            }
+        }
+    };
+
+    const handleValidation = (newValue: string) => {
+        // Should not handle validation if it is disabled
+        if (disabled) {
+            return;
+        }
+        if (validate) {
+            const error = validate(newValue) || null;
+            setErrorMessage(error);
+            if (onValidate) {
+                onValidate(error);
+            }
+        } else if (required) {
```

Comment:
Agreed, early return would be really handy here.

Would it mean that WB-1794 would be used for integrating these changes in `webapp` later? I wonder if you have considered using a snapshot release to verify if those changes will affect `webapp` right now.

Reply:
I think WB-1794 can be worked on separately since it's a pre-existing bug in both TextField and TextArea! I plan to release and integrate this current stack of PRs after this PR (it's adding up 😅). I was aiming to make only additive and backwards compatible changes to support the LabeledField work so that there's minimal changes to behaviour (with the exception to the [error initialization](https://github.com/Khan/wonder-blocks/pull/2357#discussion_r1835134581) since TextField already had this behaviour).

I ran the webapp unit tests with the snapshot of the current changes and things are passing! I'll also spot check some places manually before merging just to be sure! With WB-1794, I'd like to check more thoroughly in case there's some logic relying on the bug